6.analyze_features/scripts/9a.mean_aggregated_direct_hypothesis_test.py

```diff
@@ -0,0 +1,270 @@
+
+high_vs_wt.rename(
+    columns={
+        "levene_p_value": "high_vs_wt_levene_p_value",
+        "holm_bonferroni_p_value": "high_vs_wt_levene_p_value_holm_bonferroni",
+    },
+    inplace=True,
+)
+unsel_vs_wt.rename(
+    columns={
+        "levene_p_value": "unsel_vs_wt_levene_p_value",
+        "holm_bonferroni_p_value": "unsel_vs_wt_levene_p_value_holm_bonferroni",
+    },
+    inplace=True,
+)
+high_vs_unsel.rename(
+    columns={
+        "levene_p_value": "high_vs_unsel_levene_p_value",
+        "holm_bonferroni_p_value": "high_vs_unsel_levene_p_value_holm_bonferroni",
+    },
+    inplace=True,
+)
+
+
+# drop columns
+high_vs_wt.drop(columns=["group", "levene_statistic"], inplace=True)
+unsel_vs_wt.drop(columns=["group", "levene_statistic"], inplace=True)
+high_vs_unsel.drop(columns=["group", "levene_statistic"], inplace=True)
+
+
+# combine the data
+combined_df = pd.merge(high_vs_wt, unsel_vs_wt, on="feature")
+combined_df = pd.merge(combined_df, high_vs_unsel, on="feature")
+combined_df.head()
+
+
+# In[4]:
+
+
+variance_df_path = pathlib.Path(
+    "../../data/6.analysis_results/mean_aggregated_variance_results_each_feature.csv"
+).resolve(strict=True)
+variance_df = pd.read_csv(variance_df_path)
+variance_df.head()
+
+
+# In[5]:
+
+
+# replace metadata genotype with the correct group
+variance_df["Metadata_genotype"].replace(
+    {"high": "High-Severity", "unsel": "Mid-Severity", "wt": "Wild Type"}, inplace=True
+)
+variance_df.head()
+
+
+# In[6]:
+
+
+high_severity = variance_df.loc[variance_df["Metadata_genotype"] == "High-Severity"]
+mid_severity = variance_df.loc[variance_df["Metadata_genotype"] == "Mid-Severity"]
+wt = variance_df.loc[variance_df["Metadata_genotype"] == "Wild Type"]
+
+# rename the columns
+high_severity.rename(columns={"variance": "high_severity_variance"}, inplace=True)
+mid_severity.rename(columns={"variance": "mid_severity_variance"}, inplace=True)
+wt.rename(columns={"variance": "wild_type_severity_variance"}, inplace=True)
+
+# drop columns
+high_severity.drop(columns=["Metadata_genotype"], inplace=True)
+mid_severity.drop(columns=["Metadata_genotype"], inplace=True)
+wt.drop(columns=["Metadata_genotype"], inplace=True)
+
+high_severity.reset_index(drop=True, inplace=True)
+mid_severity.reset_index(drop=True, inplace=True)
+wt.reset_index(drop=True, inplace=True)
+
+
+# In[7]:
+
+
+all_severities = pd.merge(
+    mid_severity, high_severity, left_on="feature", right_on="feature"
+)
+all_severities = pd.merge(all_severities, wt, left_on="feature", right_on="feature")
+combined_df = pd.merge(combined_df, all_severities, on="feature")
+combined_df.head()
+
+
+# In[8]:
+
+
+# make a new column that contains permutation test results for each feature
+np.random.seed(0)
+n_permutations = 1000
+# permute each column
+permuted_df = combined_df.copy()
+for column in permuted_df.columns:
+    if column != "feature":
+        permuted_df[column] = np.random.permutation(permuted_df[column].values)
```

Comment:
oh, I see (I think) that you _are_ permuting the labels for the levene's test results? (it maybe comes down to me not fully understanding what `combined_df` is, please make sure to use descriptive variable names)

I'm not sure if this is correct...

Let's break it down again and walk through it, our specific hypothesis is that:

### The mid severity variance is significantly higher than both the low and high severity group variance _for the AreaShape feature group_

(is this correct?)

The logic then follows as:

- We want to test if we see this pattern (in AreaShape features specifically) greater than by chance. 
- We can simulate chance by randomly permuting the low/mid/high labels 1,000 times and then recalculating levene's statistic  in the same fashion we did initially (pairwise, then outputting p values for each comparison, then multiple test correction, then calculating `is_pattern_present()` [👇 low, ☝️ mid, 👇 high])
- Do this sampling _for all features_; i.e., all features get 1,000 binary scores of `is_pattern_present()`
- This gives us a distribution of how many times we see this pattern by chance _per feature_
- We can then calculate a p value per feature by counting how many times the random sampling `is_pattern_present()` count is greater than the real non-shuffled `is_pattern_present()` count and dividing by 1,000
- You can plot this p value distribution in a density curve coloring by feature group.
- Lastly, to test the AreaShape feature group specifically, we then create the 2x2 contingency table:

| | Significant p values | Non-significant p values |
| :-- | :------------- | :--------------------- |
| AreaShape group | A | B |
| Other groups | C | D |
 
- You can then apply a Fisher's exact test to this 2x2 matrix to get a single test statistic that will directly test the hypothesis above. Specifically, it will test:

> Null hypothesis: The proportion of significant p-values is the same in the AreaShape group and other groups.
> Alternative hypothesis: The AreaShape group has a higher proportion of significant p-values than the other groups.